### PR TITLE
Add payload and representations in RDF

### PR DESCRIPTION
**Pull request by csarven**
Initial requirement on Representations. This particular requirement involves issues:

* https://github.com/solid/specification/issues/45
* https://github.com/solid/specification/issues/69
* https://github.com/solid/specification/issues/109
* https://github.com/solid/specification/issues/195

Will update this space to include related issues.

Feedback on making this more readable is most welcome.

### CODE REVIEW

**Comment by bblfish** [comment]
There seems to be another question which is if a PUT of Turtle doc is done at `<foaf.ttl>`, should the `<>` refer to `<foaf.ttl>` or to the content negotiable `<foaf>`? Should clients rather PUT at `<foaf>`, and have `<foaf.ttl>` be linked to from there? The same question for jpegs and image formats and video formats.

**Comment by michielbdejong** [comment]
The way it currently works in Solid is that if you PUT turtle to `https://example.com/profile/card`, then that is its URL. You can then fetch it as Turtle or as JSON-LD or as human-viewable HTML at `https://example.com/profile/card` (or `https://example.com/profile/card#me` within that document). At no point is there any need to access the Turtle and JSON-LD (which is what we call "representation of the resource" right?) at a URL different from `https://example.com/profile/card`.

**Comment by csarven** [comment]
@bblfish I've raised something similar in https://github.com/solid/specification/issues/69#issuecomment-581340858 in context of container representations. The suggested practice for clients to not explicitly set base URI in the document, however, if set, it should refer to container's URI (as opposed to a representation URL).

>Should clients rather PUT at <foaf>, and have <foaf.ttl> be linked to from there?

Yes. Some implementations may expose representation URLs (covered in issue #109 ) but we want writes to go through the preferred content negotiable URI of the resource (ie. `<foaf>`). So, `GET /foaf` with `Accept: text/turtle` may get a response including `Content-Location: /foaf.ttl`. I've updated this PR covering what's in 109.

Note https://tools.ietf.org/html/rfc7231#appendix-B :
>   The definition of Content-Location has been changed to no longer
   affect the base URI for resolving relative URI references, due to
   poor implementation support and the undesirable effect of potentially
   breaking relative links in content-negotiated resources.
   (Section 3.1.4.2)

So, if `/foaf.ttl` includes `<>` (or equivalent in other RDF syntaxes), `GET /foaf`'s response with `Content-Location: /foaf.ttl` entails that when the document is parsed `<>` refers to URI ending `/foaf`.

Having said that, I'm not sure what servers should do, if anything, when they encounter `<>` in PUT's payload. Preferably nothing. Clients should parse the content at the resource URI instead of the representation URL to obtain the RDF graph. Is this effectively covered in the specs or do we need to mention that at all? We want to avoid exposing multiple graphs where they may only differ with references resolved against the base URI.

**Comment by csarven** [comment]
@michielbdejong Issue #109 raised whether implementations may advertise representation URLs through `Content-Location`. FWIW, servers are not forbidden from advertising representation URLs, so if some servers expose them, how should read and write operations work? The agreement was that write requests targeting the representational URL should be done through the URI of the resource.

**Comment by michielbdejong** [comment]
OK then it's #109 I disagree with :)

File: main/resource-access.bs [hunk 1]
```diff
@@ -279,6 +279,25 @@ Issue:
 Pertaining to events and loss of control mitigation:
 https://github.com/solid/specification/issues/41#issuecomment-534679278
 
+### Representations ### {#representations}
+
+When a server creates a resource on HTTP `PUT`, `POST` or `PATCH` requests
+such that the request's representation data encodes an *RDF document*
+[[!RDF11-CONCEPTS]] (as determined by the `Content-Type` header), the server
+MUST respect `GET` requests on this resource when the value of the `Accept`
```
Comment (RubenVerborgh) [inline main/resource-access.bs:287]: Yes to the idea, but the phrasing of `respect` is a bit vague/unclear to me

Comment (csarven) [reply main/resource-access.bs]: Updated to use "accept": https://github.com/solid/specification/pull/196/commits/4d8858654c4056ef53f0420cb9118f1f7ba2b9a7 ie. to honour, conform to, follow through client's wishes - matching 7231's "received, understood, and accepted" for 2xx. Will this do for now? Happy to revise later.

**Review by RubenVerborgh — APPROVED** [review]

**Review by akuckartz — APPROVED** [review]

**Comment by csarven** [comment]
@michielbdejong On the contrary, we agree on the same requirement: clients should target the resource URI for write operations.

Two approaches for Solid where both would conform to RFC 7231 (`Content-Location` is optional for servers and clients):

1. Servers and clients MUST NOT use `Content-Location`.
2. Servers MUST redirect (eg. `307`, `308` - this PR) or fail (eg. `405`) in response to write requests on representation URLs.

Solid servers MUST implement the server parts of RFC 7231 and `Content-Location` still remains optional. Disallowing `Content-Location` seems like a significant limitation (approach 1) in comparison to specifying what should happen when server uses `Content-Location` (approach 2).

Open to alternative approaches or an update to what's proposed.

File: main/resource-access.bs [hunk 2]
```diff
@@ -279,6 +279,25 @@ Issue:
...
+MUST accept `GET` requests on this resource when the value of the `Accept`
+header requests a representation in `text/turtle` or `application/ld+json`
+[[!Turtle]] [[!JSON-LD11]].
+[[Source](https://github.com/solid/specification/issues/45)]
+[[Source](https://github.com/solid/specification/issues/69)]
+[[Source](https://github.com/solid/specification/issues/109)]
+[[Source](https://github.com/solid/specification/issues/195)]
+
+When a `PUT`, `POST`, `PATCH` or `DELETE` method request targets a resource's
+representation URL, the server MUST respond with a `307` or `308` status code
```
Comment (acoburn) [inline main/resource-access.bs:296]: The redirect requirement makes sense in the case where the resource URL (e.g. `http://example.com/foo`) is different from the representation URL (e.g. `http://example.com/foo.jsonld`), but what if all representations are fetched from the same resource URL (`/foo`)? Should this sentence include something like "if the representation URL is different than the resource URL"? Or perhaps the term "representation URL" should be more rigorously defined?

Comment (csarven) [reply main/resource-access.bs]: I agree that'd be good to distinguish. Initial update here: https://github.com/solid/specification/pull/196/commits/e92d8bce5caaad42fcaad465ba326e44fc1357fc . Will this do for now? I like the idea of further clarifying how we are using "resource" and "representation" but we shouldn't define per se - perhaps just remind RFC's notions.

Comment (acoburn) [reply main/resource-access.bs]: @csarven thanks, that update works for me

**Review by acoburn — APPROVED** [review]

**Review by justinwb — APPROVED** [review]

**Comment by michielbdejong** [comment]
Yes, approach 2 is fine.

**Comment by csarven** [comment]
Great, thanks all! All approved, feedback and objections addressed.

**Pull request merged**